### PR TITLE
Add search for fields in data viewer and signals in signal viewer

### PR DESCRIPTION
**Pull request by siddhant1**

### CODE REVIEW

**Comment by domoritz** [comment]
Let's add the same for the signal viewer.

File: src/components/table/renderer.tsx [hunk 1]
```diff
@@ -14,7 +15,7 @@ const MAX_LENGTH = 150;
 export default class Table extends React.PureComponent<Props> {
   public render() {
     const headerNodes = this.props.header.map(h => (
-      <th title={h} key={h}>
+      <th onClick={() => this.props.onClickHandler && this.props.onClickHandler(h)} title={h} key={h}>
```
Comment (domoritz) [inline src/components/table/renderer.tsx]: When would the handler not be defined?

File: src/components/viz-pane/renderer.tsx [hunk 2]
```diff
@@ -57,6 +103,24 @@ export default class VizPane extends React.PureComponent<Props> {
       this.props.showLogs(true);
     }
   }
+
+  public getComponent() {
```
Comment (domoritz) [inline src/components/viz-pane/renderer.tsx]: Which component?

File: src/components/table/renderer.tsx (hunk 1 above)
Comment (siddhant1) [reply src/components/table/renderer.tsx]: When we want to reuse table without the debugger.

File: src/components/viz-pane/renderer.tsx (hunk 2 above)
Comment (siddhant1) [reply src/components/viz-pane/renderer.tsx]: I just moved the position of the function , it is used to get the data-viewer,signal-viewer and the error based on which is selected

File: src/components/table/renderer.tsx (hunk 1 above)
Comment (domoritz) [reply src/components/table/renderer.tsx]: Okay

File: src/components/viz-pane/renderer.tsx (hunk 2 above)
Comment (domoritz) [reply src/components/viz-pane/renderer.tsx]: I think it needs a qualifier in the name,

**Comment by siddhant1** [comment]
Github diff caculations is definitely struggling with this one , I moved positions of the getComponent method just because I wanted to keep it above the render method

**Comment by domoritz** [comment]
still todo: https://github.com/vega/editor/pull/437#discussion_r311030998

File: src/components/viz-pane/renderer.tsx (hunk 2 above)
Comment (siddhant1) [reply src/components/viz-pane/renderer.tsx]: What should be the name?

Comment (domoritz) [reply src/components/viz-pane/renderer.tsx]: You decide

Comment (siddhant1) [reply src/components/viz-pane/renderer.tsx]: Something like getVizPane Component will do

Comment (domoritz) [reply src/components/viz-pane/renderer.tsx]: It's not a visualization pane, right? It's a `getContextViewer` or something. Describe in the comment that it returns the debug pane, data or signal viewer.

Comment (siddhant1) [reply src/components/viz-pane/renderer.tsx]: Should I name it getContextViewer and add a comment explaining what it returns?

Comment (domoritz) [reply src/components/viz-pane/renderer.tsx]: 👍

**Comment by domoritz** [comment]
* [ ] Search in compiled Vega spec
* [ ] Try to highlight only the matching string (works in https://vega.github.io/vega-editor/)
* [ ] Try to support searching in expressions (works in https://vega.github.io/vega-editor/)
* [ ] Fix bug with missing highlight when I cycle through matches (or remove cycling)

Screen Shot 2019-08-09 at 10 37 21 AM

**Comment by domoritz** [comment]
- [ ] use correct word separators for json and add link

File: src/components/viz-pane/renderer.tsx [hunk 3]
```diff
@@ -1,46 +1,72 @@
...
-          return <DataViewer />;
-        case NAVBAR.SignalViewer:
-          return <SignalViewer />;
-        default:
-          return null;
+
+  public onClickHandler(header: string) {
+    const editor = this.props.editorRef;
+    const model = editor.getModel();
+    // wordSeparator:  /(-?\d*\.\d\w*)|([^\[\{\]\}\:\"\,\s]+)/g
```
Comment (domoritz) [inline src/components/viz-pane/renderer.tsx]: clean this up

File: src/components/viz-pane/renderer.tsx [hunk 4]
```diff
@@ -1,46 +1,72 @@
...
+    const range = model.findMatches(header, true, true, true, '/(-?d*.dw*)|([^[{]}:",s]+)/g', true)[0]
+      ? model.findMatches(header, true, true, true, '/(-?d*.dw*)|([^[{]}:",s]+)/g', true)[0].range
```
Comment (domoritz) [inline src/components/viz-pane/renderer.tsx]: Why find matches twice?

File: src/components/viz-pane/renderer.tsx [hunk 5]
```diff
@@ -57,6 +83,25 @@ export default class VizPane extends React.PureComponent<Props> {
       this.props.showLogs(true);
     }
   }
+
+  public getContextViewer() {
+    // The function returns the Component to be rendered in the Context Viewer
```
Comment (domoritz) [inline src/components/viz-pane/renderer.tsx]: Make this a proper function documentation with `/** ... */`

File: src/components/viz-pane/renderer.tsx [hunk 6]
```diff
@@ -1,46 +1,58 @@
...
-    } else {
-      return null;
+
+  public onClickHandler(header: string) {
+    const editor = this.props.editorRef;
+    const model = editor.getModel();
+    // turn on whole word search for the editor
+    !editor.getContribution('editor.contrib.findController')._state.wholeWord &&
+      editor.getContribution('editor.contrib.findController').toggleWholeWords();
+    // wordSeparator:  `~!@#$%^&*()-=+[{]}\\|;:\'",.<>/?
```
Comment (domoritz) [inline src/components/viz-pane/renderer.tsx]: Please import https://github.com/microsoft/monaco-json/blob/master/src/jsonMode.ts#L43

File: src/components/viz-pane/renderer.tsx [hunk 7]
```diff
@@ -57,6 +73,27 @@ export default class VizPane extends React.PureComponent<Props> {
       this.props.showLogs(true);
     }
   }
+
+  /**
+   *  returns the Component to be rendered in the Context Viewer
```
Comment (domoritz) [inline src/components/viz-pane/renderer.tsx]:
```suggestion
   *  Get the Component to be rendered in the Context Viewer.
```

File: src/components/input-panel/compiled-spec-display/index.tsx [hunk 8]
```diff
@@ -38,12 +43,13 @@ class CompiledSpecDisplay extends React.PureComponent<Props> {
...
           language="json"
           value={stringify(this.props.value)}
-          editorDidMount={e => {
-            e.onDidFocusEditorText(() => {
-              this.props.setEditorReference(this.refs.compiledEditor);
+          editorDidMount={editor => {
+            editor.onDidFocusEditorText(() => {
+              this.props.compiledEditorRef && this.props.compiledEditorRef.deltaDecorations(this.props.decorations, []);
+              this.props.editorRef.deltaDecorations(this.props.decorations, []);
+              localStorage.setItem('find_editor', 'compiled');
```
Comment (domoritz) [inline src/components/input-panel/compiled-spec-display/index.tsx]: Can we put that in the state?

Comment (siddhant1) [reply src/components/input-panel/compiled-spec-display/index.tsx]: Yeah , I will put that in the state!

Comment (domoritz) [reply src/components/input-panel/compiled-spec-display/index.tsx]: Thanks!

File: src/actions/editor.ts [hunk 9]
```diff
@@ -318,6 +324,15 @@ export function setEditorReference(editorRef: any) {
 
 export type SetEditorReference = ReturnType<typeof setEditorReference>;
 
+export function setCompiledEditorRef(editorRef: any) {
```
Comment (domoritz) [inline src/actions/editor.ts]: Can we not use any?

File: src/actions/editor.ts [hunk 10]
```diff
@@ -318,6 +324,15 @@ export function setEditorReference(editorRef: any) {
...
+  return {
+    editorRef: (editorRef as any).editor,
```
Comment (domoritz) [inline src/actions/editor.ts]: remove as any

File: src/constants/consts.ts [hunk 11]
```diff
@@ -57,6 +62,11 @@ export const KEYCODES = {
   SLASH: 191,
 };
 
+/**
+ * Ref : https://github.com/microsoft/monaco-editor/blob/02fb6bfd43b5ddcff36c1a1303e4aac623deff2e/monaco.d.ts#L2602
```
Comment (domoritz) [inline src/constants/consts.ts]:
```suggestion
 * Taken from: https://github.com/microsoft/monaco-editor/blob/02fb6bfd43b5ddcff36c1a1303e4aac623deff2e/monaco.d.ts#L2602
```

**Comment by siddhant1** [comment]
If we don't wrap for large tables , we will get scroll , so should we wrap or not?

File: package.json [hunk 12]
```diff
@@ -15,26 +15,29 @@
     "postcss-loader": "^3.0.0",
     "prettier": "^1.18.2",
     "style-loader": "^1.0.0",
-    "tslint-eslint-rules": "^5.4.0",
     "tslint": "^5.19.0",
+    "tslint-eslint-rules": "^5.4.0",
```
Comment (domoritz) [inline package.json]: Don't add this back

File: package.json [hunk 13]
```diff
@@ -15,26 +15,29 @@
...
+    "webpack-dev-server": "^3.8.0"
   },
   "dependencies": {
+    "@types/react": "^16.9.2",
     "@types/react-redux": "^7.1.2",
     "@types/react-router-dom": "^4.3.5",
-    "@types/react": "^16.9.2",
     "ajv": "^6.10.2",
     "d3-geo-projection": "^2.7.0",
+    "eslint": "^6.2.2",
```
Comment (domoritz) [inline package.json]: remove

File: package.json [hunk 14]
```diff
@@ -15,26 +15,29 @@
...
+    "global": "^4.4.0",
```
Comment (domoritz) [inline package.json]: what's this for?

File: src/actions/editor.ts (hunk 9 above)
Comment (domoritz) [reply src/actions/editor.ts]: Still todo

File: tslint.json [hunk 15]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (domoritz) [inline tslint.json]: don't add this back

**Pull request merged**